crossdock/src/http_server.js

```diff
@@ -0,0 +1,211 @@
+// @flow
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import bodyParser from 'body-parser';
+import ConstSampler from '../../src/samplers/const_sampler.js';
+import * as constants from './constants.js';
+import dns from 'dns';
+import express from 'express';
+import InMemoryReporter from '../../src/reporters/in_memory_reporter.js';
+import opentracing from 'opentracing';
+import request from 'request';
+import RSVP from 'rsvp';
+import Span from '../../src/span.js';
+import SpanContext from '../../src/span_context.js';
+import Tracer from '../../src/tracer.js';
+import Utils from '../../src/util.js';
+
+export default class HttpServer {
+    _tracer: Tracer;
+
+    constructor() {
+        let app = express();
+        this._tracer = new Tracer('crossdock', new InMemoryReporter(), new ConstSampler(false));
+
+        // json responses need bodyParser when working with express
+        app.use(bodyParser.json());
+
+        app.post('/start_trace', (req, res) => {
+            let startRequest = req.body;
+            console.log('start_trace_request:', startRequest);
+
+            // create or deserialize a span from the headers
```

Comment:
need to DRY this - beyond this point there's no difference in the logic between start and join, except for additional span initialization in startTrace that can be done via a lambda passed to a shared function.